doc: Governance Roles (#415)
kata-containers/community: README.md

* [Governance](#governance)
    * [Developers](#developers)
        * [Contributor](#contributor)

>         * [Maintainer](#committer)

Reviewer: ```suggestion
        * [Committer](#committer)
```

### Contributor

A Contributor to the Kata Containers project is someone who has had code merged within the last 12 months. Contributors are eligible to vote in the Architecture Committee elections. Contributors have read only access to the Kata Containers repos on GitHub.

### Committer

Kata Containers Committers (as defined by the [kata-containers-committer team](https://github.com/orgs/kata-containers/teams/kata-containers-committer))
have the ability to merge code into the Kata Containers project.
Maintainers are active Contributors and participants in the projects. In order to become a Committer, you must be nominated by established Committer and approved by quorum of the active Architecture committee via an issue against the community repo
e.g. https://github.com/kata-containers/community/issues/403. Committers have write access to the Kata Containers repos on GitHub, which
gives the ability to approve PRs, trigger the CI and merge PRs.

One of the requirements to be a committer is that you are an active contributor to the project as adjudged by the above criteria. The committer list will be required for people who no longer meet the criteria

> twice a year, lining up with the Architecture Committee election cycle. At/after the election, people who are in the kata-containers-committer team, but who haven't been an active contributor in the last 12 months will be created

Reviewer: ```suggestion
lining up with the preparation for the Architecture Committee election cycle. At that time, the kata-containers-committer team will also be reviewed, and a list of people, who are on the team but haven't been an active contributor in the last 12 months, will be created
```

> and shared with the Architecture Committee and community and after a short review period to check of errors in the tooling, will be removed from the team.

Reviewer: ```suggestion
and shared with the Architecture Committee and community. After a short review period, to allow time to check for any errors, inactive team members will be removed.
```

> [!Note]
> See [issue #413](https://github.com/kata-containers/community/issues/413) for a potential change in how active contribution is assessed.

### Admin

Kata Containers Admins (as defined by the [kata-containers-admin team](https://github.com/orgs/kata-containers/teams/kata-containers-admin) have admin access to
the kata-containers repo, allowing them to do actions like, change the branch protection rules for repositories, delete a repository and manage the access of others.

> There are sometimes good reasons to temporarily give others admin access, such as to create a secret that is used in a particular CI infrastructure, but generally

Reviewer: ```suggestion
The Admin group is intentionally kept small, however, individuals can be granted temporary admin access to carry out tasks, like creating a secret that is used in a particular CI infrastructure.
```

> There are sometimes good reasons to temporarily give others admin access, such as to create a secret that is used in a particular CI infrastructure, but generally

Reviewer 2: This part affects me the most, for instance.

I'm not part of the AC anymore, but I've been dealing -- maybe even more then AC members? -- with a lot of the admin stuff, ranging from setting up CI systems to simply unblocking PRs.

With this change, theoretically, I should have no access to do so, which will require a better velocity coming from the architecture committee.

More than that, adding someone to temporarily do something feels like a bureaucratic thing that would take more time than just having an AC member **doing** the thing. :-)

Author: > I'm not part of the AC anymore, but I've been dealing -- maybe even more than AC members? -- with a lot of the admin stuff, ranging from setting up CI systems to simply unblocking PRs.
>
> With this change, theoretically, I should have no access to do so, which will require a better velocity coming from the architecture committee.

So I think that in this circumstances you would probably fall into the 
> Optionally, some specific people that the Architecture Committee agree on adding for a specific purpose (e.g. to manage the CI)

category, but we also don't want to cover over the problem of the AC not being responsive, so should try and address that as well.

> More than that, adding someone to temporarily do something feels like a bureaucratic thing that would take more time than just having an AC member doing the thing. :-)

The idea of temporarily adding is based on things that I've had to do in the CAA repo. People setting up CI wanted to add secrets e.g. `AZURE_SUBSCRIPTION_ID` that are required in our infrastructure and whilst they could have given an admin the secret and asked them to add it, that doesn't feel like a secure process. An example of this closer to home is when you/Gaby recently added `ITA_KEY` secret (assuming this is sensitive, if not then sorry for the bad example). If you didn't have access would you (or rather Intel) have preferred to be given admin access temporarily to add that, or to send the secret to the AC for them to add (and therefore have access). I don't believe that anyone on the AC would be interested in exploiting secrets, but speaking personally, I'd rather not be involved in handling any that I don't own.

If you don't think this is a concern and others agree, then it would be more secure to block this, so I can remove it from the proposal.

Reviewer 2: > People setting up CI wanted to add secrets e.g. `AZURE_SUBSCRIPTION_ID` that are required in our infrastructure and whilst they could have given an admin the secret and asked them to add it, that doesn't feel like a secure process.

This makes sense, indeed.

> we should try and minimise this access. The admin list should be reviewed and updated after each Architecture Committee election and typically contain:

Reviewer: ```suggestion
The Admin list is reviewed and updated after each Architecture Committee election and typically contains:
```

- The Architecture Committee
- Optionally, some specific people that the Architecture Committee agree on adding for a specific purpose (e.g. to manage the CI)

### Owner

> GitHub organization owners have complete admin access to the organization, so should be limited for security reasons. The owners list should be reviewed and updated after each Architecture Committee election and contain:

Reviewer: ```suggestion
GitHub organization owners have complete admin access to the organization, and therefore this group is limited to a small number of individuals, for security reasons. The owners list is reviewed and updated after each Architecture Committee election and contains:
```

## Developers

> For contributors, there are several roles relevant to project governance:

Reviewer 3: To solidify "developer" as the umbrella term for all 4 roles:
```suggestion
For Kata developers, there are several roles relevant to project governance:
```

### Contributor

A Contributor to the Kata Containers project is someone who has had code merged within the last 12 months. Contributors are eligible to vote in the Architecture Committee elections. Contributors have read only access to the Kata Containers repos on GitHub.

### Committer

> Kata Containers Committers (as defined by the [kata-containers-committer team](https://github.com/orgs/kata-containers/teams/kata-containers-committer))

Reviewer 3: Note that this link isn't public - you have to be an org member to access it. We may or may not want to keep it here.

Author: So I think we'd also publish the proposed committers twice a year under these proposals. I think for github permissions we have to have that team as the primary source of committers though.

have the ability to merge code into the Kata Containers project.

> Committers are active Contributors and participants in the projects. In order to become a Committer, you must be nominated by established Committer and approved by quorum of the active Architecture committee via an issue against the community repo

Reviewer 3: Typos
```suggestion
Committers are active Contributors and participants in the project. In order to become a Committer, you must be nominated by an established Committer and approved by quorum of the active Architecture Committee via an issue against the community repo
```

> Committers are active Contributors and participants in the project. In order to become a Committer, you must be nominated by established Committer and approved by quorum of the active Architecture Committee via an issue against the community repo

Reviewer 3: ```suggestion
Committers are active Contributors and participants in the project. In order to become a Committer, you must be nominated by an established Committer and approved by quorum of the active Architecture Committee via an issue against the community repo
```

Author: Fixed - thanks!

but who haven't been an active contributor in the last 12 months will be created and shared with the Architecture Committee and community.
After a short review period, to allow time to check for any errors, inactive team members will be removed.

> [!Note]
> See [issue #413](https://github.com/kata-containers/community/issues/413) for a potential change in how active contribution is assessed.

### Admin

Kata Containers Admins (as defined by the [kata-containers-admin team](https://github.com/orgs/kata-containers/teams/kata-containers-admin) have admin access to
the kata-containers repo, allowing them to do actions like, change the branch protection rules for repositories, delete a repository and manage the access of others.
The Admin group is intentionally kept small, however, individuals can be granted temporary admin access to carry out tasks, like creating a secret that is used in a particular CI infrastructure.
The Admin list is reviewed and updated after each Architecture Committee election and typically contains:
- The Architecture Committee

> - Optionally, some specific people that the Architecture Committee agree on adding for a specific purpose (e.g. to manage the CI)

Reviewer 2: I understand we're taking this a security measure to prevent something to happen.  However, the community has always been based on trust and work being done.

Having this feels like we're shooting ourselves in the foot, as either you're an AC member, member of OIF, or you depend on their own criteria (regardless of them being actually active or not) to decide who'll be actually helping the project.

Now, if we want to go through a formal process, I'd **strongly** recommend that any change in the admin side would to go through a vote and the majority of the admins would have to agree before any change happens (including secrets being added, machines being added, PRs being merged, etc).  The malicious actor may come from anywhere, including AC / OIF.

Reviewer 3: IMO this language is fine since it mostly codifies what is already roughly the current state of things. 

> Now, if we want to go through a formal process, I'd strongly recommend that any change in the admin side would have to go through a vote and the majority of the admins would have to agree before it happens (including secrets being added, machines being added, PRs being merged, etc). The malicious actor may come from anywhere, including AC / OIF.

I suggest discussing adding this clause in another PR/issue since this PR has already been up for so long.

Reviewer 4: > I understand we're taking a security measure to prevent something to happen. However, the community has always been based on trust and work being done.

I don't think the proposed text hampers trust on any way. People will still manage to gain admin powers if this is beneficial for the project. The concern is more around the housekeeping when people leave the project IMHO.

### Owner

GitHub organization owners have complete admin access to the organization, and therefore this group is limited to a small number of individuals, for security reasons.
The owners list is reviewed and updated after each Architecture Committee election and contains:
- The Community Manager and one, or more extra people from the `OpenInfra Foundation` for redundancy and vacation cover
- The Architecture Committee
- Optionally, some specific people that the Architecture Committee agree on adding for a specific purpose (e.g. to help with repo/CI migration)

## Architecture Committee

> The Architecture Committee is responsible for architectural decisions, including standardization, and making final decisions if Committers disagree. It is comprised of 7 members, who are elected by contributors.

Reviewer 3: In general in this doc I would use:
 * "Contributor" (capitalized) to refer to the specific Contributor developer role.
 * "developer" to refer broadly to the 4 developer roles.

```suggestion
The Architecture Committee is responsible for architectural decisions, including standardization, and making final decisions if Committers disagree. It is comprised of 7 members, who are elected by Contributors.
```

Author: I think I've updated based on your suggestions...
